src/main/scala/chisel3/util/Counter.scala

```diff
@@ -19,27 +19,47 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   *     ...
   *   }
   * }}}
-  *
-  * @param n number of counts before the counter resets (or one more than the
-  * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
+  require(r.start >= 0 && r.end >= 0, s"Counter range must be positive, got: $r")
+
+  private val delta = math.abs(r.step).U
+  private val width = math.max(log2Up(r.start), log2Up(r.end)) + 1
```

Comment:
```suggestion
  private val width = math.max(log2Up(r.head), log2Up(r.last)) + 1
```
I think using the actual first and last (rather than specified `start` and `end`) actually brings the step and inclusivity into the calculation. Here's a Scastie I used to decide what to use:
https://scastie.scala-lang.org/LsvLdMkoQBeTcXUoZh4brA

Reply:
Thanks, that is much better :+1: